text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1101 @@
+trait into a nonconst trait and a const trait and "merge" them by making one of them be a super trait of the other:
+
+```rust
+const trait Foo {
+    fn foo(&self);
+}
+trait Bar: Foo {
+    fn bar(&self);
+}
+
+impl const Foo for () {
+    fn foo(&self) {}
+}
+impl Bar for () {
+    fn bar(&self) {
+        println!("writing to terminal is not possible in const eval");
+    }
+}
+```
+
+Such a split is not possible without a breaking change, so splitting may not be feasible in some cases.
+Especially since we may later offer the ability to have const and nonconst methods on the same trait, then allowing
+the traits to be merged again. That's churn we'd like to avoid.
+
+Note that it may frequently be that such a trait should have been split even without constness being part of the picture.
+
+Similarly one may want an always-const method on a trait of otherwise non-const methods:
+
+```rust
+const trait InitFoo {
+    fn init(i: i32) -> Self;
+}
+trait Foo: const InitFoo {
+    const INIT: Self = Self::init(0);
+    fn do_stuff(&mut self);
+}
+```
+
+or even only offer `INIT` if `InitFoo` is `const`:
+
+```rust
+const trait InitFoo: Sized {
+    fn init(i: i32) -> Self;
+}
+trait Foo: InitFoo {
+    const INIT: Self = Self::init(0) where Self: const InitFoo;
+    fn do_stuff(&mut self);
+}
+```
+
+# Alternatives
+[alternatives]: #alternatives
+
+## What is the impact of not doing this?
+
+We would require everything that wants a const-equivalent to have duplicated traits and not
+use `const` fn at all, but use associated consts instead. Similarly this would likely forbid
+invoking builtin operators. This same concern had been brought up for the `const fn` stabilization
+[7 years ago](https://github.com/rust-lang/rust/issues/24111#issuecomment-385046163).
+
+Basically what we can do is create
+
+```rust
+trait ConstDefault {
+    const DEFAULT: Self;
+}
+```
+
+and require users to use
+
+```rust
+const FOO: Vec<u8> = ConstDefault::DEFAULT;
+```
+
+instead of
+
+```rust
+const fn FOO: Vec<u8> = Default::default();
+```
+
+This duplication is what this RFC is suggesting to avoid.
+
+Since it has already been possible to do all of this on stable Rust for years, and no major
+crates have popped and gotten used widely, I assume that is either because
+
+* it's too much duplication, or
+* everyone was waiting for the work (that this RFC wants to stabilize) to finish, or
+* both.
+
+So while it is entirely possible that rejecting this RFC and deciding not to go down this route
+will lead to an ecosystem for const operations to be created, it would result in duplication and
+inconsistencies that we'd rather like to avoid.
+
+Such an ecosystem would also make `const fn` obsolete, as every `const fn` can in theory be represented
+as a trait, it would just be very different to use from normal rust code and not really allow nice abstractions to be built.
```

Comment:
`const` generics has limitations, so no, it would not make `const fn` obsolete…

Reply:
"Redundant" may be the better word. And yes, we'd need more const generics features